marlowe-runtime/web-server/Language/Marlowe/Runtime/Web/Server/REST/Contracts.hs

```diff
@@ -160,18 +199,67 @@ put
     Nothing -> throwError $ notFound' "Contract not found"
     Just (Left (TempTx _ Unsigned Tx.ContractCreated{txBody})) -> do
-      textEnvelope <- fromDTOThrow (badRequest' "Invalid body value") body
-      addField ev $ Body textEnvelope
-      tx <- either (const $ throwError $ badRequest' "Invalid body text envelope content") pure $ deserialiseFromTextEnvelope (AsTx AsBabbage) textEnvelope
-      unless (getTxBody tx == txBody) $ throwError (badRequest' "Provided transaction body differs from the original one")
+      let
+        -- `<|>` gives me error here
+        req :: Maybe (Either (Cardano.Tx BabbageEra) (ShelleyTxWitness BabbageEra))
+        req = case fromDTO (Left body) of
+          Just res -> pure res
+          Nothing -> fromDTO (Right body)
+
+      for_ (fromDTO body :: Maybe Cardano.TextEnvelope) \te ->
+        addField ev $ Body te
+
+      tx <- case req of
+        Nothing -> throwError $ badRequest' "Invalid body value"
+        Just (Left tx) -> do
+          unless (getTxBody tx == txBody) $ throwError (badRequest' "Provided transaction body differs from the original one")
+          pure tx
+        -- It seems that wallets provide nearly empty `TxWitness` back. Here is a quoat from `CIP-30` docs:
+        --
+        -- > Only the portions of the witness set that were signed as a result of this call are returned to
+        -- > encourage dApps to verify the contents returned by this endpoint while building the final transaction.
+        Just (Right (ShelleyTxWitness (TxWitness wtKeys _ _ _ _))) -> do
+          let
+            txScriptValidityToScriptValidity :: TxScriptValidity era -> ScriptValidity
+            txScriptValidityToScriptValidity TxScriptValidityNone = ScriptValid
+            txScriptValidityToScriptValidity (TxScriptValidity _ scriptValidity) = scriptValidity
+
+            scriptValidityToIsValid :: ScriptValidity -> Alonzo.IsValid
+            scriptValidityToIsValid ScriptInvalid = Alonzo.IsValid False
+            scriptValidityToIsValid ScriptValid = Alonzo.IsValid True
+
+            txScriptValidityToIsValid :: TxScriptValidity era -> Alonzo.IsValid
+            txScriptValidityToIsValid = scriptValidityToIsValid . txScriptValidityToScriptValidity
+
+            tx = case (txBody, makeSignedTransaction [] txBody) of
+              (ShelleyTxBody era txBody' _ _ txmetadata scriptValidity, ShelleyTx _ (ValidatedTx _ bkTxWitness _ _)) -> do
+                let
+                  TxWitness _ bkBoot bkScripts bkDats bkRdmrs = bkTxWitness
+                  wt' =
+                    TxWitness
+                      wtKeys
+                      bkBoot
+                      bkScripts
+                      bkDats
+                      bkRdmrs
+
+                ShelleyTx era $ ValidatedTx
+                  txBody'
+                  wt'
+                  (txScriptValidityToIsValid scriptValidity)
+                  (maybeToStrictMaybe txmetadata)
+          pure tx
```

Comment:
What a pain... I really want to be able to remove all this submit stuff from the runtime.